<launch>
    <!-- This launch file runs the pose-based visual servoing -->

    <!-- Robot namespace -->
    <arg name="robot_name" default="my_gen3"/>
    <arg name="start_delay_seconds" default="0"/>

    <!--
    ### Original Param Values ###
    control_hz = 60
    control_gain = 0.75
    control_deadband_error = 0.0002
    xyz_vel_limit = 0.18
    rpy_vel_limit = 0.18
    target_camera_frame = desired_cam_frame
    current_camera_frame = camera_color_frame
    robot_base_frame = base_link
    command_ee_cartesian_velocity_topic = control_input
    -->

    <!-- Control Related -->
    <arg name="control_hz"                                  default="60.0" />   <!--default: 60.0-->
    <arg name="control_gain"                                default=".75" />    <!--default: .75-->
    <arg name="control_deadband_error"                      default="0.002" />  <!--default: 0.0002-->
    
    <!-- Robot Cart Vel Limit -->
    <arg name="xyz_vel_limit"                               default="0.15" />    <!--default: 0.18-->
    <arg name="rpy_vel_limit"                               default="0.15" />    <!--default: 0.18-->

    <!-- Frames for Visual Servoing -->
    <arg name="target_camera_frame"                         default="desired_cam_frame" />
    <arg name="current_camera_frame"                        default="camera_color_frame" />
    <arg name="robot_base_frame"                            default="base_link" />

    <!-- Control Command Topic -->
    <arg name="command_ee_cartesian_velocity_topic"         default="control_input" />

    <!-- PBVS Node -->
    <node name="pbvs" pkg="raf" type="pbvs" output="screen" launch-prefix="bash -c 'sleep $(arg start_delay_seconds); $0 $@' ">
        <param name="robot_name" value="$(arg robot_name)"/>

        <param name="pbvs_control_loop_hz"              type="double" value="$(arg control_hz)" />
        <param name="pbvs_control_law_gain_lambda"      type="double" value="$(arg control_gain)" />
        <param name="pbvs_control_deadband_error"       type="double" value="$(arg control_deadband_error)" />

        <param name="xyz_vel_limit"                     type="double" value="$(arg xyz_vel_limit)" />
        <param name="rpy_vel_limit"                     type="double" value="$(arg rpy_vel_limit)" />

        <param name="desired_camera_frame"              type="string" value="$(arg target_camera_frame)" />
        <param name="current_camera_frame"              type="string" value="$(arg current_camera_frame)" />
        <param name="robot_base_frame"                  type="string" value="$(arg robot_base_frame)" />

        <param name="control_input_topic"               type="string" value="$(arg command_ee_cartesian_velocity_topic)" />
        
    </node>

</launch>